reconcile/terraform_cloudflare_dns.py

```diff
@@ -0,0 +1,246 @@
+            f"account {cf_acct.name} because it does define a terraform state "
+        )
+
+    integrations = tf_state.integrations or []
+    for i in integrations or []:
+        name = i.integration
+        if name.replace("-", "_") == QONTRACT_INTEGRATION:
+            # we have to ensure the bucket key(file) is unique across
+            # Cloudflare accounts to support running per-account
+            bucket_key = f"{QONTRACT_INTEGRATION}-{cf_acct.name}-{zone}.tfstate"
+            bucket_name = tf_state.bucket
+            bucket_region = tf_state.region
+            break
+
+    if bucket_name and bucket_key and bucket_region:
+        backend_config = TerraformS3BackendConfig(
+            aws_acct_creds["aws_access_key_id"],
+            aws_acct_creds["aws_secret_access_key"],
+            bucket_name,
+            bucket_key,
+            bucket_region,
+        )
+    else:
+        raise ValueError(f"No state bucket config found for account {aws_acct.name}")
+
+    return backend_config
+
+
+def get_cf_acct_config(
+    cf_acct: CloudflareAccountV1, secret_reader: SecretReader
+) -> CloudflareAccountConfig:
+    cf_acct_creds = secret_reader.read_all({"path": cf_acct.api_credentials.path})
+    if not cf_acct_creds.get("api_token") or not cf_acct_creds.get("account_id"):
+        raise SecretIncompleteError(
+            f"secret {cf_acct.api_credentials.path} incomplete: api_token and/or account_id missing"
+        )
+    cf_acct_config = CloudflareAccountConfig(
+        cf_acct.name,
+        cf_acct_creds["api_token"],
+        cf_acct_creds["account_id"],
+        cf_acct.enforce_twofactor or DEFAULT_CLOUDFLARE_ACCOUNT_2FA,
+        cf_acct.q_type or DEFAULT_CLOUDFLARE_ACCOUNT_TYPE,
+    )
+    return cf_acct_config
+
+
+def build_clients(
+    secret_reader: SecretReader,
+    query_zones: CloudflareDnsZoneQueryData,
+    selected_account: Optional[str] = None,
+) -> list[tuple[str, TerrascriptCloudflareClient]]:
+    clients = []
+    cf_acct_configs = {}
+    for zone in query_zones.zones or []:
+        cf_acct = zone.account
+        cf_acct_name = cf_acct.name
+        if selected_account and cf_acct_name != selected_account:
+            continue
+        if cf_acct_name in cf_acct_configs:
+            cf_acct_config = cf_acct_configs[cf_acct_name]
+        else:
+            cf_acct_config = get_cf_acct_config(cf_acct, secret_reader)
+            cf_acct_configs[cf_acct_name] = cf_acct_config
+        aws_acct = cf_acct.terraform_state_account
+        aws_backend_config = create_backend_config(
+            secret_reader, aws_acct, cf_acct, zone.identifier
+        )
+
+        ts_config = create_cloudflare_terrascript(
+            cf_acct_config,
+            aws_backend_config,
+            cf_acct.provider_version,
+        )
+
+        ts_client = TerrascriptCloudflareClient(ts_config)
+        clients.append((cf_acct.name, ts_client))
+    return clients
+
+
+def cloudflare_dns_zone_to_external_resource(
```

Comment:
I'd definitely include a good docstring when you get around to it explaining why this is happening.